File: python/src/agent_framework/core/in_memory_memory.py
from __future__ import annotations

from datetime import datetime, timezone


class InMemoryMemoryProvider:
    def __init__(self) -> None:
        self._store: dict[str, dict] = {}

    def _key(self, scope: str, key: str) -> str:
        return f"{scope}:{key}"

    async def get(self, key: str, scope: str = "working") -> str | None:
        entry = self._store.get(self._key(scope, key))
        return None if entry is None else entry["value"]

    async def set(self, key: str, value: str, scope: str = "working") -> None:
        self._store[self._key(scope, key)] = {
            "key": key,
            "value": value,
            "scope": scope,
            "updated_at": datetime.now(timezone.utc).isoformat(),
        }

    async def search(self, query: str, limit: int = 10) -> list[dict[str, str]]:
        results = [entry for entry in self._store.values() if _matches_memory_query(entry, query)]
        return sorted(results, key=lambda item: item["key"])[:limit]


def _matches_memory_query(entry: dict[str, str], query: str) -> bool:
    needle = query.lower()
    key = entry["key"].lower()
    value = entry["value"].lower()
    if needle in key or needle in value or (key and key in needle):
        return True
    tokens = [token for token in needle.split() if len(token) > 2]
    return any(token in key or token in value for token in tokens)

File: typescript/packages/core/src/in-memory-memory.ts
import type { MemoryEntry, MemoryProvider } from './types/contracts.js';

export class InMemoryMemoryProvider implements MemoryProvider {
  private store = new Map<string, MemoryEntry>();

  private key(scope: 'working' | 'persistent', key: string): string {
    return `${scope}:${key}`;
  }

  async get(key: string, scope: 'working' | 'persistent' = 'working'): Promise<string | null> {
    return this.store.get(this.key(scope, key))?.value ?? null;
  }

  async set(key: string, value: string, scope: 'working' | 'persistent' = 'working'): Promise<void> {
    this.store.set(this.key(scope, key), {
      key,
      value,
      scope,
      updatedAt: new Date().toISOString(),
    });
  }

  async search(query: string, limit = 10): Promise<MemoryEntry[]> {
    return [...this.store.values()]
      .filter((entry) => matchesMemoryQuery(entry, query))
      .sort((a, b) => a.key.localeCompare(b.key))
      .slice(0, limit);
  }

  async clear(scope?: 'working' | 'persistent'): Promise<void> {
    if (!scope) {
      this.store.clear();
      return;
    }
    for (const [composite, entry] of this.store.entries()) {
      if (entry.scope === scope) {
        this.store.delete(composite);
      }
    }
  }
}


function matchesMemoryQuery(entry: MemoryEntry, query: string): boolean {
  const needle = query.toLowerCase();
  const key = entry.key.toLowerCase();
  const value = entry.value.toLowerCase();
  if (key.includes(needle) || value.includes(needle) || (key.length > 0 && needle.includes(key))) {
    return true;
  }
  const tokens = needle.split(/\s+/).filter((token) => token.length > 2);
  return tokens.some((token) => key.includes(token) || value.includes(token));
}
